### ISSUE TITLE
Slim base path strips leading slash

### ISSUE THREAD

**Issue by x00x70** [userfrosting/UserFrosting#775]
https://github.com/slimphp/Slim/issues/1724

Causing issues with CSRF Blacklist.

**Comment by alexweissman** [comment]
What happened to the discussion we had about this?  Simply saying "causing issues" makes it difficult for me to pick up where we left off.

**Comment by x00x70** [comment]
sorry dude, been completely slammed.  Think we decided we were going to strip leading slashes from the paths when checking against csrf blacklist, so the regex in the config file doesn't need a leading slash.

**Comment by alexweissman** [comment]
Well, I do like the idea of making the blacklist routes somewhat consistent with the route definition API.  In that case we'd actually normalize paths so that they _do_ have a leading `/`.

**Comment by alexweissman** [comment]
what happens if someone wants to blacklist the `/` route?  Would the regex be `^$`?

**Comment by x00x70** [comment]
Ah, yeah - i meant we'd need to add a trailing slash to routes like the assets.  Not sure how we proceed. i am completely awash lately and haven't been able to work on UF stuff in a few weeks. likely to stay this way for another month.

**Comment by alexweissman** [comment]
I've made some discoveries:

- Starting in Slim 4, `Uri::getPath()` will [always strip the leading slash](https://github.com/slimphp/Slim/pull/2115)
- Laravel's equivalent concept [does not use leading slashes](https://laravel.com/docs/5.4/csrf#csrf-excluding-uris).  Then again, neither does Laravel's route definitions, so in that case the route paths and blacklist paths are consistent.

If we go with "always enforce a leading slash", we'll basically be prepending a `/` to all paths returned by Slim 4's `Uri::getPath()`.  I still think this would be worth it, in the interest of remaining consistent with the route definition syntax.  In that case, your blacklists would look like:

```
        '^/api/hoots' => [
            'POST'
        ],
        '^/api/screeches' => [
            'POST',
            'PUT'
        ]
```

We could even consider eliminating the need for `^`.  Are there cases where you might want to match a route somewhere in the middle?  Seems possible but unlikely.

**Comment by x00x70** [comment]
with regards to leading `^` - better to have it and not need it then need it and not be able to, matching in the middle of a route is quite conceivable.
I agree with the rest about always having a leading slash.

**Comment by alexweissman** [comment]
Alright, I made my decision to require a leading slash in the blacklisted routes.  Implemented as of v4.1.8.
